crates/chain/src/indexed_tx_graph.rs

```diff
@@ -271,42 +271,35 @@ where
         block: &Block,
         height: u32,
     ) -> ChangeSet<A, I::ChangeSet> {
-        let block_id = BlockId {
-            hash: block.block_hash(),
-            height,
-        };
         let mut changeset = ChangeSet::<A, I::ChangeSet>::default();
         for (tx_pos, tx) in block.txdata.iter().enumerate() {
             changeset.indexer.merge(self.index.index_tx(tx));
             if self.index.is_tx_relevant(tx) {
                 let txid = tx.compute_txid();
-                let anchor = A::from_block_position(block, block_id, tx_pos);
                 changeset.tx_graph.merge(self.graph.insert_tx(tx.clone()));
-                changeset
-                    .tx_graph
-                    .merge(self.graph.insert_anchor(txid, anchor));
+                changeset.tx_graph.merge(
+                    self.graph
+                        .insert_anchor(txid, TxPosInBlock::new(block, height, tx_pos).into()),
+                );
             }
         }
         changeset
     }
 
     /// Batch insert all transactions of the given `block` of `height`.
     ///
-    /// Each inserted transaction's anchor will be constructed from
-    /// [`AnchorFromBlockPosition::from_block_position`].
+    /// Each inserted transaction's anchor will be constructed using [`TxPosInBlock`].
     ///
     /// To only insert relevant transactions, use [`apply_block_relevant`] instead.
     ///
     /// [`apply_block_relevant`]: IndexedTxGraph::apply_block_relevant
-    pub fn apply_block(&mut self, block: Block, height: u32) -> ChangeSet<A, I::ChangeSet> {
-        let block_id = BlockId {
-            hash: block.block_hash(),
-            height,
-        };
+    pub fn apply_block(&mut self, block: &Block, height: u32) -> ChangeSet<A, I::ChangeSet> {
         let mut graph = tx_graph::ChangeSet::default();
         for (tx_pos, tx) in block.txdata.iter().enumerate() {
-            let anchor = A::from_block_position(&block, block_id, tx_pos);
-            graph.merge(self.graph.insert_anchor(tx.compute_txid(), anchor));
+            graph.merge(self.graph.insert_anchor(
+                tx.compute_txid(),
+                TxPosInBlock::new(block, height, tx_pos).into(),
+            ));
```

Comment:
This means we are constructing `TxPosInBlock` on every tx, thus re-computing the block hash for every tx... I have a couple of ideas to fix this.

1. Make all fields of `TxPositionInBlock` public and get rid of construction and getter methods.
2. Add a method to change the `tx_pos`. I.e. `fn with_tx_pos(self, tx_pos: usize) -> Self`.

I'm in favor of 1. because it is simpler and I don't see much value in having an API that enforces the consistency of the type. I don't see the caller constructing this type. The main usecase of this type (as I understand) is for implementing `From<TxPositionInBlock>` for anchors types that support it.

Reply:
> Make all fields of TxPositionInBlock public and get rid of construction and getter methods.

Made fields public as suggested.

Although that puts us back [to the beginning](https://github.com/bitcoindevkit/bdk/pull/1594#discussion_r1751085427) (suggestion to pass only height, not BlockId), the repeated calculation of block hash should be avoided. Most likely why the creation of `BlockId` was in the original code in a first place.